pkg/controllers/user/approuter/nginx_controller.go

```diff
@@ -0,0 +1,130 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rancher/rancher/pkg/ticker"
+
+	workloadutil "github.com/rancher/rancher/pkg/controllers/user/workload"
+	"github.com/rancher/rancher/pkg/settings"
+	rkek8s "github.com/rancher/rke/k8s"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+var (
+	nginxIngressLabels = map[string]string{
+		"app": "ingress-nginx",
+	}
+	renewInterval = 5 * time.Minute
+)
+
+type NginxIngressController struct {
+	workloadController workloadutil.CommonController
+	nodeLister         v1.NodeLister
+	podLister          v1.PodLister
+	dnsClient          *Client
+	clusterName        string
+}
+
+func (n *NginxIngressController) sync(key string, obj *workloadutil.Workload) error {
+	rootDomain := settings.RootDomain.Get()
+	serverURL := settings.BaseRDNSServerURL.Get()
+	if rootDomain == "" || serverURL == "" { //rootDomain and serverURL need to be set when enable rdns controller
+		return nil
+	}
+
+	if !labels.SelectorFromSet(nginxIngressLabels).Matches(labels.Set(obj.Labels)) {
+		return nil
+	}
+
+	ips, err := n.getNginxControllerIPs()
+	if err != nil {
+		return err
+	}
+	if serverURL != n.dnsClient.GetBaseURL() {
+		n.dnsClient.SetBaseURL(serverURL)
+	}
+
+	fqdn := n.getRootFqdn(rootDomain)
+	if err := n.dnsClient.ApplyDomain(fqdn, ips); err != nil {
+		logrus.WithError(err).Errorf("update fqdn [%s] to server [%s] error", fqdn, serverURL)
+		return err
+	}
+	logrus.Infof("update fqdn [%s] to server [%s] success", fqdn, serverURL)
+	return nil
+}
+
+func (n *NginxIngressController) getNginxControllerIPs() ([]string, error) {
+	var ips []string
+	pods, err := n.podLister.List(defaultNginxIngressNamespace, labels.SelectorFromSet(labels.Set(nginxIngressLabels)))
+	if err != nil {
+		logrus.WithError(err).Error("syncing ingress rules to rdns server error")
+		return nil, err
+	}
+	for _, pod := range pods {
+		ip, err := n.getNodePublicIP(pod.Spec.NodeName)
+		if err != nil {
+			logrus.WithError(err).Errorf("get node %s public ip error", pod.Spec.NodeName)
+			continue
+		}
+		ips = append(ips, ip)
+	}
+	return ips, nil
+}
+
+func (n *NginxIngressController) getNodePublicIP(nodeName string) (string, error) {
+	node, err := n.nodeLister.Get("", nodeName)
+	if err != nil {
+		return "", err
+	}
+	//from ips
+	ip := ""
+	for _, address := range node.Status.Addresses {
+		if address.Type == "ExternalIP" {
+			return address.Address, nil
+		}
+		if address.Type == "InternalIP" {
+			ip = address.Address
+		}
+	}
+
+	//from annotation
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+
+	if ip, ok := node.Annotations[rkek8s.ExternalAddressAnnotation]; ok {
+		return ip, nil
+	}
+
+	if ip, ok := node.Annotations[rkek8s.InternalAddressAnnotation]; ok {
+		return ip, nil
+	}
+
+	return ip, nil
+}
+
+func (n *NginxIngressController) getRootFqdn(rootDomain string) string {
+	return fmt.Sprintf("%s.%s", n.clusterName, rootDomain)
```

Comment:
The `cluster_name` may be duplicated for different users.
We must get something unique like cluster_uuid, not sure if there is a cluster_uuid.

Reply:
Check the API: `https://xxxx/v3/clusters`
Can we use the UUID of the cluster object? To use the first eight is fine.

Reply:
The DNS service must return a random domain.  The client can't choose a domain.